README.md
[![New Relic One Catalog Project header](https://github.com/newrelic/opensource-website/raw/master/src/images/categories/New_Relic_One_Catalog_Project.png)](https://opensource.newrelic.com/oss-category/#new-relic-one-catalog-project)

[![Pull Request](https://github.com/newrelic/nr1-pathpoint/actions/workflows/pr.yml/badge.svg)](https://github.com/newrelic/nr1-pathpoint/actions/workflows/pr.yml)
[![Node.js CI](https://github.com/newrelic/nr1-pathpoint/actions/workflows/node.js.yml/badge.svg)](https://github.com/newrelic/nr1-pathpoint/actions/workflows/node.js.yml)
[![Repolinter Action](https://github.com/newrelic/nr1-pathpoint/actions/workflows/repolinter.yml/badge.svg?branch=main)](https://github.com/newrelic/nr1-pathpoint/actions/workflows/repolinter.yml)
[![NR1 Library Deprecation Checks](https://github.com/newrelic/nr1-pathpoint/actions/workflows/nr1_library_deprecation_check.yml/badge.svg)](https://github.com/newrelic/nr1-pathpoint/actions/workflows/nr1_library_deprecation_check.yml)
[![Release](https://github.com/newrelic/nr1-pathpoint/actions/workflows/release.yml/badge.svg)](https://github.com/newrelic/nr1-pathpoint/actions/workflows/release.yml)




# Pathpoint

## NEW RELIC STAFF: For a working e-commerce config example compatible with Demotron use [this one](examples/nr-demotron-order-processing.json)

Other useful resources:

- [Pathpoint Demo Script](https://docs.google.com/document/d/1j1vO4J1RD77-VyJBwMrgMv9YfjjM7zHjkKENO34ryO4/edit?usp=sharing)
- [Dictionary of useful Pathpoint terms](https://docs.google.com/document/d/1eYuMxQelrmitJ5_SO74gut_jqTokFCi33Py77s1mK0c/edit?usp=sharing)



### Pathpoint now has a new GUI editor for stages, steps, and touchpoints!  Please provide feedback to jhagan@newrelic.com.

## Other Pathpoint Config Examples:
- *Dummy* config (suitable for any NR account)](examples/Pathpoint_Json_v1.5%20(13).json)
- [Simple e-commerce example](examples/e-commerce-membership.json)
- [video/media process example](examples/video-processing.json)

![UI Overview](screenshots/main-ui.png)

## Description

Pathpoint is an enterprise platform tracker that models system health in relation to actual user-impacting business stages.  

### Stages

These are the highest level business stages.   At the actual system level these will be a rollup of many services and methods.  PathPoint will give us a view of Latency, Utilization and Errors for each high level business stage.  As an example from an e-commerce customer may identify the following business critical stages: PRE-PROCESSING, PRE-ORDER, CHECKOUT, ORDER MANAGEMENT, DISTRIBUTION & RETURNS.

![Stages](screenshots/stages.png)

### Steps

These are “sub-stages” of a parent stage.  that represent a more granular aggregation of services.  Clicking on a step will highly show more detailed services and functions in the TouchPoints list below.  If a stage has a red border it means there is an error anomaly for this stage.  Below are example steps from an ecommerce company.  In this example these steps correspond to APM services.

![Steps](screenshots/steps.png)

### Touchpoints

These are the most granular entities in the PathPoint model.  PathPoint TouchPoints are often a specific APM or Browser application, but you can use any NRQL query as a touchpoint.  The health status of a TouchPoint will be tied to error rate and latency.

![Touchpoints](screenshots/touchpoints.png)

### Canary Filter

![Canary Filter](screenshots/canary.png)

### Flame Filter

![Canary Filter](screenshots/flame.png)

## Open source license

This project is distributed under the [Apache 2 license](LICENSE).

## What do you need to make this work?

Pathpoint will work in nearly any New Relic account.  To get started you'll need some telemetry in your account that you care about.  This could be any of the following telemetry types: Metrics, Events, Logs, Traces.  A common starting place for Pathpoint is APM Events and Logs, but it is up to you.   The other thing you'll need to know is how this telemetry maps onto the business process you want to model as stages and steps.  That may require some internal disucssions with your stakeholders to understand how things really fit  in.

When you are ready to make some edits you can simply download the current version of the JSON config, edit and re-upload it.  You can also use "right click" to adjust touchpoint configurations.

## In-Product Support

In the Pathpoint UI it is possible to submit a support issue which will be triaged by the Pathpoint team.   They will respond within a few days to update you on our overall assesment and if we are able to resolve the issue we'll provide an ETA.

## Getting started

First, ensure that you have [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) and [NPM](https://www.npmjs.com/get-npm) installed. If you're unsure whether you have one or both of them installed, run the following command(s) (If you have them installed these commands will return a version number, if not, the commands won't be recognized):

```bash
git --version
npm -v
```

Next, install the [NR1 CLI](https://one.newrelic.com/launcher/developer-center.launcher) by going to [this link](https://one.newrelic.com/launcher/developer-center.launcher) and following the instructions (5 minutes or less) to install and setup your New Relic development environment.

Next, to clone this repository and run the code locally against your New Relic data, execute the following command:

```bash
nr1 nerdpack:clone -r https://github.com/newrelic/nr1-pathpoint.git
cd nr1-pathpoint
nr1 nerdpack:serve
```

Visit [https://one.newrelic.com/?nerdpacks=local](https://one.newrelic.com/?nerdpacks=local), navigate to the Nerdpack, and :sparkles:


## Deploying this Nerdpack

Open a command prompt in the nerdpack's directory and run the following commands.

```bash
# To create a new uuid for the nerdpack so that you can deploy it to your account:
nr1 nerdpack:uuid -g [--profile=your_profile_name]

# To see a list of APIkeys / profiles available in your development environment:
# nr1 profiles:list
nr1 nerdpack:publish [--profile=your_profile_name]
nr1 nerdpack:deploy [-c [DEV|BETA|STABLE]] [--profile=your_profile_name]
nr1 nerdpack:subscribe [-c [DEV|BETA|STABLE]] [--profile=your_profile_name]
```

Visit [https://one.newrelic.com](https://one.newrelic.com), navigate to the Nerdpack, and :sparkles:

## Enabling the Support System

In order for the support system to work you will need to configure a Jira endpoint to submit issues to a special Jira project.  The variable that governs this is `proxyJira`.  This value may be pre-populated in the NR1 catalog repository, but if not you will need to get the valid URL from your New Relic account team.

```bash
{
    "proxyJira":"[Pathpoint Proxy URL]"
}
```

## Support

New Relic has open-sourced this project. This project is provided AS-IS WITHOUT WARRANTY OR DEDICATED SUPPORT. Issues and contributions should be reported to the project here on GitHub.

We encourage you to bring your experiences and questions to the [Explorers Hub](https://discuss.newrelic.com) where our community members collaborate on solutions and new ideas.

### Community

New Relic hosts and moderates an online forum where customers can interact with New Relic employees as well as other customers to get help and share best practices. Like all official New Relic open source projects, there's a related Community topic in the New Relic Explorers Hub. You can find this project's topic/threads here:

https://discuss.newrelic.com/t/pathpoint

### Issues / enhancement requests

Issues and enhancement requests can be submitted in the [Issues tab of this repository](../../issues). Please search for and review the existing open issues before submitting a new issue.

### Security

As noted in our [security policy](https://github.com/newrelic/nr1-pathponit/security/policy), New Relic is committed to the privacy and security of our customers and their data. We believe that providing coordinated disclosure by security researchers and engaging with the security community are important means to achieve our security goals.

If you believe you have found a security vulnerability in this project or any of New Relic's products or websites, we welcome and greatly appreciate you reporting it to New Relic through [HackerOne](https://hackerone.com/newrelic).

## Contributing

Contributions are encouraged! If you submit an enhancement request, we'll invite you to contribute the change yourself. Please review our [Contributors Guide](CONTRIBUTING.md).

Keep in mind that when you submit your pull request, you'll need to sign the CLA via the click-through using CLA-Assistant. If you'd like to execute our corporate CLA, or if you have any questions, please drop us an email at opensource+nr1-pathpoint@newrelic.com.


docs/CHANGELOG.md
## [1.7.6](https://github.com/newrelic/nr1-pathpoint/compare/v1.7.5...v1.7.6) (2022-12-21)


### Bug Fixes

* Fixed the regex to increase the measuring time range ([#55](https://github.com/newrelic/nr1-pathpoint/issues/55)) ([865e627](https://github.com/newrelic/nr1-pathpoint/commit/865e6278f43037fdf8e0c62c9b8d7717269f1069))

## [1.7.5](https://github.com/newrelic/nr1-pathpoint/compare/v1.7.4...v1.7.5) (2022-12-15)


### Bug Fixes

* Fixed time picker for WLD touchpoints ([#54](https://github.com/newrelic/nr1-pathpoint/issues/54)) ([a352818](https://github.com/newrelic/nr1-pathpoint/commit/a352818d537bc105415f6edc252bcb2d00b5b6a1))

## [1.7.4](https://github.com/newrelic/nr1-pathpoint/compare/v1.7.3...v1.7.4) (2022-12-15)


### Bug Fixes

*  Resolve issues on Touchpoint GUI editor & time picker ([#53](https://github.com/newrelic/nr1-pathpoint/issues/53)) ([15c0d26](https://github.com/newrelic/nr1-pathpoint/commit/15c0d264edf10d286b674f095937cf1e0bf0593c))

## [1.7.3](https://github.com/newrelic/nr1-pathpoint/compare/v1.7.2...v1.7.3) (2022-11-10)


### Bug Fixes

* Added detailed install instructions from app catalog ([#50](https://github.com/newrelic/nr1-pathpoint/issues/50)) ([5d399eb](https://github.com/newrelic/nr1-pathpoint/commit/5d399eb7d654048cba303f08834de392d9570848))

## [1.7.2](https://github.com/newrelic/nr1-pathpoint/compare/v1.7.1...v1.7.2) (2022-07-21)


### Bug Fixes

* Update config file version to 1.7.1 ([3c12011](https://github.com/newrelic/nr1-pathpoint/commit/3c12011e9151ea01895f83495965ce06ef6574e0))

## [1.7.1](https://github.com/newrelic/nr1-pathpoint/compare/v1.7.0...v1.7.1) (2022-07-18)


### Bug Fixes

* Remove release notes in favor of change log. ([9b0bc83](https://github.com/newrelic/nr1-pathpoint/commit/9b0bc830636aebc72286dc826457992ebd5b6b5e))

# [1.7.0](https://github.com/newrelic/nr1-pathpoint/compare/v1.6.3...v1.7.0) (2022-06-17)


### Features

* Steps document Update ([#48](https://github.com/newrelic/nr1-pathpoint/issues/48)) ([7b82fb2](https://github.com/newrelic/nr1-pathpoint/commit/7b82fb2138de5a831e1740e0c0ffb5be5fd5228b))

## [1.6.3](https://github.com/newrelic/nr1-pathpoint/compare/v1.6.2...v1.6.3) (2022-06-01)


### Bug Fixes

* Removed emulation flag ([7133ee4](https://github.com/newrelic/nr1-pathpoint/commit/7133ee4e3a1f7c469c267cb147c16842a5700c07))

## [1.6.2](https://github.com/newrelic/nr1-pathpoint/compare/v1.6.1...v1.6.2) (2022-05-31)


### Bug Fixes

* Updated image assets and fonts ([f233053](https://github.com/newrelic/nr1-pathpoint/commit/f233053ab985de3fb4abd396a230a93274bf5a96))

## [1.6.1](https://github.com/newrelic/nr1-pathpoint/compare/v1.6.0...v1.6.1) (2022-05-20)


### Bug Fixes

* Bug when running in emulator mode ([#44](https://github.com/newrelic/nr1-pathpoint/issues/44)) ([8ba187d](https://github.com/newrelic/nr1-pathpoint/commit/8ba187dafa5113f83e0b923f97a4db4544837295))

# [1.6.0](https://github.com/newrelic/nr1-pathpoint/compare/v1.5.8...v1.6.0) (2022-05-09)


### Features

* New GUI editor for stages, steps, and touchpoints ([a28eff9](https://github.com/newrelic/nr1-pathpoint/commit/a28eff9a1dcc06758e0b329c8388e989319d5ffc))

## [1.5.6](https://github.com/newrelic/nr1-pathpoint/compare/v1.5.5...v1.5.6) (2022-02-25)


### Bug Fixes

* app deprecations ([#38](https://github.com/newrelic/nr1-pathpoint/issues/38)) ([d457943](https://github.com/newrelic/nr1-pathpoint/commit/d457943dd52e30a77e1163954fa9114ad155044d))

## [1.5.5](https://github.com/newrelic/nr1-pathpoint/compare/v1.5.4...v1.5.5) (2022-01-27)


### Bug Fixes

* Bump version ([749de3a](https://github.com/newrelic/nr1-pathpoint/commit/749de3a93236adb378fcb5382d018341ca243b4b))
* empty commit to bump version ([#34](https://github.com/newrelic/nr1-pathpoint/issues/34)) ([febc46d](https://github.com/newrelic/nr1-pathpoint/commit/febc46d1d9002772bf782b0e9901d5272779dbb0))
* release-bump ([e2fbade](https://github.com/newrelic/nr1-pathpoint/commit/e2fbadecb4cd0d9a45f5be0ab30c6c2eab8d7805))

# 1.0.0 (2022-01-27)


### Bug Fixes

* **catalog:** Merge pull request [#9](https://github.com/newrelic/nr1-pathpoint/issues/9) from jbeveland27/chore/catalog ([1a2fbd1](https://github.com/newrelic/nr1-pathpoint/commit/1a2fbd16c40c9c8ab630970cfa1f5de5956d32fb))
* empty commit to bump version ([#34](https://github.com/newrelic/nr1-pathpoint/issues/34)) ([febc46d](https://github.com/newrelic/nr1-pathpoint/commit/febc46d1d9002772bf782b0e9901d5272779dbb0))
* force release ([93b7593](https://github.com/newrelic/nr1-pathpoint/commit/93b75936244c333db6b9f2bb06db27bcc508177f))
* release-bump ([e2fbade](https://github.com/newrelic/nr1-pathpoint/commit/e2fbadecb4cd0d9a45f5be0ab30c6c2eab8d7805))
* trigger build ([591c230](https://github.com/newrelic/nr1-pathpoint/commit/591c230ef7e5e33e9dbf6efc0443c9e57cfb56b1))
* updating release version ([cefcd04](https://github.com/newrelic/nr1-pathpoint/commit/cefcd04490845b30a96f00ca3fc2edcf0656f79e))


### Features

* add catalog keywords and categoryTerms ([ebced7d](https://github.com/newrelic/nr1-pathpoint/commit/ebced7d6d32892fab60e5910bfe6e66cc3c02396))
* update semantic release with node ([#13](https://github.com/newrelic/nr1-pathpoint/issues/13)) ([91a8aec](https://github.com/newrelic/nr1-pathpoint/commit/91a8aec9456c2a55b73de033e4f078156665f3df))


### Reverts

* Revert "Configuration Jest" ([30a778d](https://github.com/newrelic/nr1-pathpoint/commit/30a778dbaeee327060b993a9b406557d2ddca882))

# 1.0.0 (2022-01-27)


### Bug Fixes

* **catalog:** Merge pull request [#9](https://github.com/newrelic/nr1-pathpoint/issues/9) from jbeveland27/chore/catalog ([1a2fbd1](https://github.com/newrelic/nr1-pathpoint/commit/1a2fbd16c40c9c8ab630970cfa1f5de5956d32fb))
* empty commit to bump version ([#34](https://github.com/newrelic/nr1-pathpoint/issues/34)) ([febc46d](https://github.com/newrelic/nr1-pathpoint/commit/febc46d1d9002772bf782b0e9901d5272779dbb0))

setup/import_touchpoints.php
<?php
// Enable Dashboards creation
//------------------------------------------------------------------------------------
$st = file_get_contents("../.env.json");
$config = json_decode($st, true);
$APIKEY = "API_KEY";//$config["API_KEY"];
$accountID = "ACCOUNT_ID";//$config["accountID"];
require("newrelicFunctions.php");
//------------------------------------------------------------------------------------

$f = fopen("touchpoints.csv", "r");
$stage_names = [];

$string = file_get_contents("../nerdlets/pathpoint-nerdlet/config/view.json");
$view = json_decode($string, true);

for ($i = 0; $i < count($view["stages"]); $i++) {
    $stage_names[] = $view["stages"][$i]["title"];
}

$touchPoints = [
    [
        "index" => 0,
        "country" => "PRODUCTION",
        "touchpoints" => []
    ]
];
removeTouchpoints($view);
$createDashboards = false;
//----------------------------------------------------------------------------
$message   =  "\n\nDo you whant to create default Dashboards on the client Account?[y/n]";
print $message;
flush();
$confirmation  =  trim( fgets( STDIN ) );
if ( $confirmation == 'y' ) {
   $createDashboards = true;
   print("\nCreateting Dashboards...\n\n");
   $actualDashboardsList = getDashboardsList('pathpoint--', $APIKEY);
}

$xx = 0;
$readingHeader = true;
while (($datos = fgetcsv($f, 1000, ",")) !== FALSE) {
    if ($readingHeader) {
        $readingHeader = false;
    } else {
        $stages[0] = $datos[0];
        $touchpoint = $datos[1];
        $steps = $datos[2];
        $touchpointType = $datos[3];
        $min_count = $datos[4];
        $min_apdex = $datos[5];
        $max_response_time = $datos[6];
        $max_error_percentage = $datos[7];
        $max_avg_response_time = $datos[8];
        $max_total_check_time = $datos[9];
        $min_success_percentage = $datos[10];
        $multiAccountID = $datos[11];
        $tp_query = $datos[12];
        if($createDashboards){
            $touchpointLink = createDashboardForStage($stages[0], $touchpoint, $actualDashboardsList);
        }else{
            $touchpointLink = $datos[13];
        }
        $measureTime = $datos[14];
        $touchpoint_timeout = $datos[15];
        
        //----------------------------------
        foreach ($stages as $stage) {
            if (valida_stage($stage)) {
                addTouchpoint(stage_index($stage), $touchpointType, $touchpoint, $steps, $touchpointLink, $tp_query, $min_count, $min_apdex, $max_response_time, $max_error_percentage, $max_avg_response_time, $max_total_check_time, $min_success_percentage,$multiAccountID, $measureTime, $touchpoint_timeout);
            }
        }
    }
}
fclose($f);

$ff = fopen("temporal", "w");
fwrite($ff, json_encode($touchPoints));
fclose($ff);
shell_exec("cat temporal | jq > ../nerdlets/pathpoint-nerdlet/config/touchPoints.json");

$ff = fopen("temporal", "w");
fwrite($ff, json_encode($view));
fclose($ff);
shell_exec("cat temporal | jq > ../nerdlets/pathpoint-nerdlet/config/view.json");


function valida_stage($stage)
{
    global $stage_names;
    foreach ($stage_names as $stname) {
        if ($stname == $stage) {
            return true;
        }
    }
    return false;
}

function stage_index($stage)
{
    global $stage_names;
    for ($index = 0; $index < count($stage_names); $index++) {
        if ($stage_names[$index] == $stage) {
            return $index;
        }
    }
}

function removeTouchpoints(&$view)
{
    global $stage_names;
    for ($i = 0; $i < count($stage_names); $i++) {
        $view["stages"][$i]["touchpoints"] = [];
        for ($ss = 0; $ss < count($view["stages"][$i]["steps"]); $ss++) {
            if (array_key_exists("sub_steps", $view["stages"][$i]["steps"][$ss])) {
                for ($s = 0; $s < count($view["stages"][$i]["steps"][$ss]["sub_steps"]); $s++) {
                    $view["stages"][$i]["steps"][$ss]["sub_steps"][$s]["relationship_touchpoints"] = [];
                }
            } else {
                $view["stages"][$i]["steps"][$ss]["relationship_touchpoints"] = [];
            }
        }
    }
}

function getStepIndex($stage_index, $steps)
{
    global $view;
    $s = explode(",", $steps);

    $result = [];
    if( $steps == ''){
        return $result;
    }
    foreach ($s as $code) {
        $code = trim($code);
        //print("stage-".$stage_index." code:".$code."\n");
        $result[] = $view["stages"][$stage_index]["steps"][$code[0] - 1]["sub_steps"][ord($code[1]) - 65]["index"];
    }
    return $result;
}

function setRelationship($stage_index, $relationIndex, $tpIndex)
{
    global $view;
    for ($ss = 0; $ss < count($view["stages"][$stage_index]["steps"]); $ss++) {
        if (array_key_exists("sub_steps", $view["stages"][$stage_index]["steps"][$ss])) {
            for ($s = 0; $s < count($view["stages"][$stage_index]["steps"][$ss]["sub_steps"]); $s++) {
                if ($view["stages"][$stage_index]["steps"][$ss]["sub_steps"][$s]["index"] == $relationIndex) {
                    $view["stages"][$stage_index]["steps"][$ss]["sub_steps"][$s]["relationship_touchpoints"][] = $tpIndex;
                }
            }
        } else {
            if ($view["stages"][$stage_index]["steps"][$ss]["index"] == $relationIndex) {
                $view["stages"][$stage_index]["steps"][$ss]["relationship_touchpoints"][] = $tpIndex;
            }
        }
    }
}

function getDashboardLink($link)
{
    if ($link != '') {
        return $link;
    } else {
        return false;
    }
}

function addTouchpoint($stage_index, $touchpointType, $touchpoint, $steps, $touchpointLink, $tp_query, $min_count, $min_apdex, $max_response_time, $max_error_percentage, $max_avg_response_time, $max_total_check_time, $min_success_percentage, $multiAccountID, $measureTime, $touchpoint_timeout)
{
    global $view;
    global $touchPoints;
    $tpIndex = count($view["stages"][$stage_index]["touchpoints"]) + 1;
    $historicError = false;
    $touchPoints[0]["touchpoints"][] = [
        "stage_index" => ($stage_index + 1),
        "value" => $touchpoint,
        "touchpoint_index" => $tpIndex,
        "status_on_off" => true,
        "relation_steps" => getStepIndex($stage_index, $steps),
        "measure_points" => []
    ];
    $last = count($touchPoints[0]["touchpoints"]);
    if ($touchpointType == 'PRC') {
        $touchPoints[0]["touchpoints"][$last - 1]["measure_points"][] = [
            "type" => $touchpointType,
            "timeout" => (int) $touchpoint_timeout,
            "query" => $tp_query,
            "min_count" => (int) $min_count,
            "max_count" => (int) $min_count+100,
            "session_count" => 0
        ];
    }else if ($touchpointType == 'PCC') {
        $touchPoints[0]["touchpoints"][$last - 1]["measure_points"][] = [
            "type" => $touchpointType,
            "timeout" => (int) $touchpoint_timeout,
            "query" => $tp_query,
            "min_count" => (int) $min_count,
            "max_count" => (int) $min_count+100,
            "transaction_count" => 0
        ];
    }else if ($touchpointType == 'APP') {
        $touchPoints[0]["touchpoints"][$last - 1]["measure_points"][] = [
            "type" => $touchpointType,
            "timeout" => (int) $touchpoint_timeout,
            "query" => $tp_query,
            "min_apdex" => (float) $min_apdex,
            "max_response_time" => (float) $max_response_time,
            "max_error_percentage" => (float) $max_error_percentage,
            "apdex_value" => 0,
            "response_value" => 0,
            "error_percentage" => 0
        ];
    }else if ($touchpointType == 'FRT') {
        $touchPoints[0]["touchpoints"][$last - 1]["measure_points"][] = [
            "type" => $touchpointType,
            "timeout" => (int) $touchpoint_timeout,
            "query" => $tp_query,
            "min_apdex" => (float) $min_apdex,
            "max_response_time" => (float) $max_response_time,
            "max_error_percentage" => (float) $max_error_percentage,
            "apdex_value" => 0,
            "response_value" => 0,
            "error_percentage" => 0
        ];
    }else if ($touchpointType == 'SYN') {
        $touchPoints[0]["touchpoints"][$last - 1]["measure_points"][] = [
            "type" => $touchpointType,
            "timeout" => (int) $touchpoint_timeout,
            "query" => $tp_query,
            "max_avg_response_time" => (float) $max_avg_response_time,
            "max_total_check_time" => (float) $max_total_check_time,
            "min_success_percentage" => (float) $min_success_percentage,
            "success_percentage" => 0,
            "max_duration" => 0,
            "max_request_time" =>0
        ];
    }else if ($touchpointType == 'WLD') {
        $touchPoints[0]["touchpoints"][$last - 1]["measure_points"][] = [
            "type" => $touchpointType,
            "timeout" => (int) $touchpoint_timeout,
            "query" => $tp_query,
            "status_value" => 'NO-VALUE'
        ];
    }else if ($touchpointType == 'DRP') {
        $touchPoints[0]["touchpoints"][$last - 1]["measure_points"][] = [
            "type" => $touchpointType,
            "timeout" => (int) $touchpoint_timeout,
            "query" => $tp_query,
            "value" => 0
        ];
    }
    if($multiAccountID !== ''){
        $lastMeasure = count($touchPoints[0]["touchpoints"][$last - 1]["measure_points"]);
        $touchPoints[0]["touchpoints"][$last - 1]["measure_points"][$lastMeasure-1]["accountID"] = (int) $multiAccountID;
    }
    if($measureTime !== ''){
        $lastMeasure = count($touchPoints[0]["touchpoints"][$last - 1]["measure_points"]);
        $touchPoints[0]["touchpoints"][$last - 1]["measure_points"][$lastMeasure-1]["measure_time"] = $measureTime;
    }
    $view["stages"][$stage_index]["touchpoints"][] = [
        "index" => $tpIndex,
        "stage_index" => $stage_index + 1,
        "status_on_off" => true,
        "response_error" => false,
        "show_grey_square" => false,
        "active" => false,
        "value" => $touchpoint,
        "highlighted" => false,
        "error" => false,
        "history_error" => $historicError,
        "countrys" => [
            0
        ],
        "dashboard_url" => [
            getDashboardLink($touchpointLink)
        ],
        "relation_steps" => getStepIndex($stage_index, $steps)
    ];
    foreach (getStepIndex($stage_index, $steps) as $relationIndex) {
        setRelationship($stage_index, $relationIndex, $tpIndex);
    }
}

function createDashboardForStage($stage, $touchpoint, $actualDashboardsList)
{
    global $APIKEY;
    global $accountID;
    //print_r($actualDashboardsList["dashboards"]); exit;
    $dashboardTitle = "pathpoint-" . $stage . "-" . $touchpoint;
    foreach ($actualDashboardsList["dashboards"] as $dashboard) {
        //print_r($dashboard); exit;
        if ($dashboard["title"] == $dashboardTitle) {
            // print("Removing Existing DashboardID=" . $dashboard['id'] . " Name:" . $dashboard['title']);
            // $result = deleteDashboardByID($dashboard['id'], $APIKEY);
            // if ($result["dashboard"]["id"] == $dashboard['id']) {
            //     print("--->OK\n");
            // } else {
            //     print("-->ERROR\n");
            // }

            // Do not create Dashboard because it already exists
            print("Dashboard allready exist--> ID=" . $dashboard['id'] . "\n");
            $dashboardGuid = getDashboardGuid($dashboardTitle, $APIKEY, $accountID);
            print("DashbordGUID=" . $dashboardGuid . "\n");
            return "https://one.newrelic.com/redirect/entity/" . $dashboardGuid;
        }
    }
    // Create a New Dashboars
    $string = file_get_contents("dashboard_model.json");
    $dashboardModel = json_decode($string, true);
    //Update the Dashdoard model Account ID
    for($i=0;$i<count($dashboardModel["dashboard"]["widgets"]);$i++){
        $dashboardModel["dashboard"]["widgets"][$i]["account_id"]= intval($accountID);
        //print($dashboardModel["dashboard"]["widgets"][$i]["account_id"]."\n");
    }
    $apmServiceGuid = getApmServiceGuid($touchpoint, $APIKEY, $accountID);

    $dashboardModel["dashboard"]["title"] = $dashboardTitle;
    $apmLink = "";
    if ($apmServiceGuid) {
        $apmLink = "https://one.newrelic.com/redirect/entity/" . $apmServiceGuid;
    }
    $dashboardModel["dashboard"]["widgets"][0]["data"][0]["source"] = "[Go to APM Dashboard](" . $apmLink . ")\n";
    $dashboardModel["dashboard"]["widgets"][1]["data"][0]["nrql"] = "SELECT count(*),max(duration) as max,min(duration) as min,average(duration) as average,median(duration),percentage(count(*), WHERE error is true) as percentage_error" .
        " FROM Transaction WHERE appName='" . $touchpoint . "'" .
        " SINCE 30 minutes ago";
    $dashboardModel["dashboard"]["widgets"][1]["presentation"]["title"] = $touchpoint . " Statistics";

    $dashboardModel["dashboard"]["widgets"][2]["data"][0]["nrql"] = "SELECT count(*) as count" .
        " FROM Transaction WHERE appName='" . $touchpoint . "'" .
        " TIMESERIES auto SINCE 30 minutes ago";

    $dashboardModel["dashboard"]["widgets"][3]["data"][0]["nrql"] = "SELECT average(duration),median(duration),max(duration),min(duration) " .
        " FROM Transaction WHERE appName='" . $touchpoint . "'" .
        " TIMESERIES AUTO SINCE 30 minutes ago";

    $dashboardModel["dashboard"]["widgets"][4]["data"][0]["nrql"] = "SELECT count(*)" .
        " FROM Transaction WHERE appName='" . $touchpoint . "'" .
        " TIMESERIES AUTO FACET request.headers.userAgent SINCE 30 minutes ago";

    $dashboardModel["dashboard"]["widgets"][5]["data"][0]["nrql"] = "SELECT count(*)" .
        " FROM Transaction WHERE appName='" . $touchpoint . "'" .
        " FACET request.headers.host  SINCE 30 minutes ago";

    $dashboardId = createDashboard(json_encode($dashboardModel), $APIKEY);
    print("Created Dashboard--> ID=" . $dashboardId . "\n");

    $dashboardGuid = getDashboardGuid($dashboardTitle, $APIKEY, $accountID);
    print("DashbordGUID=" . $dashboardGuid . "\n");
    return "https://one.newrelic.com/redirect/entity/" . $dashboardGuid;
}
